### PR TITLE
Fix undefined forward referenced ID in decorations

### PR DESCRIPTION
**Pull request by Qining**
Fix issue #185 by dumping unreachable blocks (if they are terminated by
branch/return instructions).

Deadcode elimination will be implemented later to remove the unreachable code and related IDs.

### CODE REVIEW

**Comment by johnkslang** [comment]
This is a pretty big change to make to the emitted code.  There was some driving reason the dead code was eliminated originally, do we know what that was, and why it is now safe to emit it?

I actually think the smaller perturbation to make is to eliminate the bad decorations, and leave the generated code (which is now being translated by real drivers) the way it currently is.

**Comment by Qining** [comment]
Hi John, @dekimir and I think this is a safe and quick fix of the issue #185 and enable using validator by default in Vulkan CTS.  But eventually we will have deadcode elimination.

File: SPIRV/spvIR.h [hunk 1]
```diff
@@ -287,7 +288,17 @@ class Function {
             parameterInstructions[p]->dump(out);
 
         // Blocks
-        inReadableOrder(blocks[0], [&out](const Block* b) { b->dump(out); });
+        // Dump the reachable blocks in readable order first.
+        // Unreachable blocks which are terminated with branch/return
+        // instructions will be dumped after.
+        std::unordered_set<const Block*> dumped;
+        inReadableOrder(blocks[0], [&out, &dumped](const Block* b) {
+          b->dump(out);
```
Comment (dekimir) [inline SPIRV/spvIR.h]: Indentation incosistent with the surrounding code.

**Comment by dekimir** [comment]
IIRC, we started skipping all unreachable blocks in `dump()` in #135 because it was easier to implement the readable-order traversal that way.  I hadn't anticipated issues like #185.

@johnkslang if you'd rather we eliminated the expression decorations dangling due to non-emitted blocks, we can do that instead.  It'll just take a bit more time, delaying slightly the glslang-output validation in the Vulkan CTS.

**Comment by johnkslang** [comment]
> if you'd rather we eliminated the expression decorations dangling due to non-emitted blocks, we can do that instead

Yes, I think that's the better thing to do, for a couple of reasons.  Thanks!

**Comment by dekimir** [comment]
@johnkslang Just a clarification: do we want this dead-code elimination to happen

- before `dump()`, operating on `Block` and `Instruction` objects, or
- during `dump()`, as clever skipping of some `Instruction`s, or
- after `dump()` as a SPIRV-only optimization pass?

**Comment by johnkslang** [comment]
This is a bit of a special case, because the bug exists, but I don't want to give the extra dead code to the drivers, so whatever is fastest to get us to: the only observed change is the bad decorations disappear.

I suspect that's your first choice:  go delete the dead code right before dumping, and take eliminate the dead decorations at the same time.

This could be later followed up by a more complete DCE pass, or even starting to look at how we want to structure optimization/transform passes, etc. as part of glslang, or independent tools, etc.  But, like I said in the 1st paragraph, this one feels like a special case to me that doesn't need to wait for this bigger answer.

**Comment by Qining** [comment]
Close as the issue is resolved in 5639f3aca5b75cbe5419a623eecf5e3794fab917

**Pull request closed without merging**